src/Prima/Pyxis/Form/Autocomplete.elm

```diff
@@ -92,51 +111,161 @@ The configuration of the `Autocomplete`'s `State`.
 type alias StateConfig =
     { focused : Maybe String
     , selected : Maybe String
-    , filter : Maybe String
+    , filter : Filter
     , isMenuOpen : Bool
-    , choices : List AutocompleteChoice
+    , choices : ChoicesStatus
+    , quiteFor100ms : Debouncer.Debouncer Msg Msg
+    , threshold : Int
     }
 
 
 {-| Initializes the `Autocomplete`'s `State`.
 -}
-init : List AutocompleteChoice -> State
-init choices =
-    State <| StateConfig Nothing Nothing Nothing False choices
+init : State
+init =
+    State <| StateConfig Nothing Nothing Nothing False Pristine (initDebouncer 0) 1
+
+
+{-| Internal. Debouncer initializer
+-}
+initDebouncer : Float -> Debouncer.Debouncer o o
+initDebouncer secondsDebounce =
+    Debouncer.manual
+        |> Debouncer.settleWhenQuietFor (Just <| Debouncer.fromSeconds secondsDebounce)
+        |> Debouncer.toDebouncer
 
 
 {-| Updates the `Autocomplete`'s `State`.
 -}
-update : Msg -> State -> State
+update : Msg -> State -> ( State, Cmd Msg, Filter )
```

Comment:
rimango molto perplesso da questa forma che costringe lato consumo a salvarsi un Autocomplete come tripletta nel modello (oppure a svrapparsi obbligatoriamente questa cosa nel ciclo di update anche se un consumatore non è interessato al debouncing e alla fetch from server). Sarebbe anche l'unico componente ad avere questo modo di aggiornarsi. Inoltre ad elm non piacciono le tuple con più di due elementi.

Ti propongo di rendere il debounce configurabile e splittare l'update in uno che aggiorna e uno che emette side effects così per non spaccare anche i casi che attualmente non lo usano. Esponendo il filter con un metodo pick su State in modo da permettere lato consumo di decidere quando lanciare il fetch (es se è lungo più di n caratteri).

`
withDebounce: Float -> (Msg -> msg) -> State -> State
-- Modifier che salva nella configurazione che vuoi attivare il debouncer (che a quel punto è maybe, di default a Nothing. Se uno lo vuole lo configura nella fase init dell'autocomplete) e le due funzioni di mapping che ti permettono di risolvere l'astrazione con il model e msg del consumatore.
`

`update: Msg -> State -> State
-- quasi come ora nel senso che aggiorna Input Filter e il Debouncer nello stato interno ma non emette side effects
`

`debounce: Msg -> State -> Cmd msg
-- Se il debounce è stato configurato (se non lo è salame chi usa l'API emetti Cmd none) emetti i tuoi side effect.
Se msg è input resetti debouncer, 
Se è Debounce vai a prenderti il valore aggiornato di Debouncer.update che hai prodotto in update ed emetti l'evento di aggiornamento del filtro se ha emesso oppure riemetti un nuovo Debounce. 

Questa funzione viene raccolta in un withCmd dal consumatore in modo pulito ed emessa senza problemi (il messaggio puoi già wrapparlo tu col mapper ricevuto con withDebounce
`

Reply:
Capisco quello che dici e mentre lo integravo sia per Domus che per l'esempio, ho avuto la tentazione di creare una funzione per estrarre solo lo State, ma ogni volta mi sono reso conto che non era necessario. Il consumer non deve salvarsi la tripletta, difatti l'init restituisce solo lo State. Essendo un autocomplete, quella struttura è la più ideale per l'utilizzo in uno scenario reale. Se qualcuno non vuole gestirsi l'aggiornamento delle opzioni, vuol dire che probabilmente avrebbe dovuto usare il componente Select. Mi è stato detto che attualmente nessuno usa questo Autocomplete, quindi prima di aggiungere altra sovrastruttura per permettere uno scenario che non si verificherà potenzialmente mai, vorrei ridiscutere la cosa insieme anche ad Alessio e Matteo.

Reply:
Forse @Gsciarra non ti ho dato tutte le informazioni, colpa mia 😅 
In pratica l'autocomplete viene utilizzato per esigenze di design anche con un set statico di dati. In alcuni casi infatti è niente di più che una select con un filtro (che poi è la nostra definizione di autocomplete). 

Il fatto di utilizzare il debouncer ci sta ma userei una piccola accortezza per non obbligare il consumatore a scrivere codice non necessario in nessuno dei due use-cases (cioè lista di opzioni proveniente da endpoint oppure lista di opzioni statiche). 

Quindi sarei per avere *due diverse funzioni di update*: la prima, quella standard, è quella attuale `Msg -> State -> State`, la seconda invece è un `updateWithDebouncer` (nome a caso) la quale restituisce un eventuale `Cmd` da lanciare. 

In questo modo nel nostro caso possiamo usarla com'è attualmente, lasciando però aperta la possibilità ad altri di inserirvi una chiamata http/graphql.

Reply:
Tra l'altro così facendo la funzione `updateWithDebouncer` potrebbe internamente chiamare la `update` normale e poi aggiungere la propria logica di gestione del debouncer

Reply:
il consumer dovrà o salvarsi la tripletta o scrivere una funzione che swrappa la tripletta ed estragga State (e salvarselo), Filter e il Cmd per andare poi a consumarli in qualche modo andando a sporcare il ciclo di update con funzioni di appoggio. Cosa che in realtà non è necessaria perché Filter fa già parte dello State ed è sempre interrogabile una volta aggiornato per sapere se effettuare o meno la chiamata.
Nei cicli di update generalmente hai questa situazione (preso da example di Form del pacchetto)

```elm
model
                |> updateAutocomplete subMsg
                |> H.withoutCmds
```
dove updateAutocomplete va a riaggiornare  lo State dell'autocomplete all'interno del tuo modello.
Scrivendo un update non omomorfco mi obblighi a dichiarare l'updating di Autocomplete in una funzione locale dentro a un `let in` e poi passare i tuoi tre valori alle mie funzioni che già sono presenti. è un approccio che stiamo cercando di evitare in tutti i modi scrivendo cicli di update più fluidi possibile.

```elm
let
       (state, cmd, filter) =  Autocomplete.update subMsg model.myAutocomplete 
in
model
     |> updateAutocomplete state
     |> withCmds 
          [ cmd
          , if (filter > 3) then 
                  Task.fetchMyData 
            else 
                  Cmd.none
          ]

Usando invece l'altro approccio con le utility di mapping verrebbe

```elm
model
    |> (.myAutocomplete >> Autocomplete.update subMsg)
    |> withCmdsMap 
        [ .myAutocomplete >> Autocomplete.debounce subMsg
        , ifThenCmdMap (Model.isFilterGreaterThan 3)
             Task.fetchMyData
        ]
```

Reply:
io invece sono per usare un update solo, perchè avere due funzioni diverse non ti impedisce di fare un autocomplete server e poi usare update semplice. Inoltre il discorso "non può andare in pipe" non è un motivo valido per cambiare una api. La pipeline non è così idiomatica in elm da guidare una api, e si risolve tutto con una funzioncina di helper che si scrive in due righe.